docs/decisions/0023-handlebars-template-engine.md

```diff
@@ -0,0 +1,176 @@
+  - Using the delimiter "`-`" for helpers registered to handle the kernel functions, to distinguish them from each other and from our system or built-in Handlebars helpers.
+- Supporting both positional and hash arguments, for passing parameters to the helpers, and validating the arguments for the required type and count.
+- Handling the output types, formats, and errors of the helpers, including complex types or JSON schemas.
+- Implementing the helpers in a performant and secure way, and avoiding any side effects or unwanted modifications to the template context or data.
+
+Effectively, there will be four buckets of helpers enabled in the Handlebars Template Engine:
+
+1. Default helpers from the Handlebars library to enable loops and conditions (#if, #each, #with, #unless)
+2. Functions in the kernel
+3. Helpers helpful to prompt engineers (i.e., message, or)
+4. Utility helpers that can be used to perform simple logic or transformations on the template data or arguments (i.e., set, get, json, concat, equals, range, array)
+
+**HandlebarsPromptTemplate pseudocode**
+
+A prototype implementation of a handlebars prompt template engine with built-in helpers could look something like this:
+
+```csharp
+public async Task<string> RenderAsync(Kernel kernel, ContextVariables contextVariables, CancellationToken cancellationToken = default)
+{
+  return RenderAsync(kernel, contextVariables, new Dictionary<string, object?>(), cancellationToken);
+}
+
+// Overloaded method to support a dict of objects as template variables
+public async Task<string> RenderAsync(Kernel kernel, ContextVariables contextVariables, Dictionary<string, object?> templateVariables, CancellationToken cancellationToken = default)
+{
+  var handlebars = HandlebarsDotNet.Handlebars.Create();
+
+  RegisterKernelFunctionsAsHelpers(kernel, contextVariables, handlebars, templateVariables);
+
+  RegisterSystemHelpers(handlebars, templateVariables);
+
+  var template = handlebars.Compile(this._promptModel.Template);
+
+  var prompt = template(templateVariables);
+
+  return await Task.FromResult(prompt).ConfigureAwait(true);
+}
+
+private static void RegisterKernelFunctionsAsHelpers(
+  Kernel kernel,
+  ContextVariables contextVariables,
+  IHandlebars handlebarsInstance,
+  Dictionary<string, object?> templateVariables,
+  CancellationToken cancellationToken = default)
+{
+  foreach (IKernelPlugin plugin in kernel.Plugins)
+  {
+      foreach (KernelFunction function in plugin)
+      {
+        handlebarsInstance.RegisterHelper($"{plugin.Name}-{function.Name}", (in HelperOptions options, in Context context, in Arguments arguments)) =>
+        {
+          // 1. Get parameters from HB template arguments
+          // 2. Port parameter values to context variables
+          // 3. Invoke kernel function and write result to the template
+        }
+      }
+  }
+
+}
+
+private static void RegisterSystemHelpers(
```

Comment:
We should provide a way when constructing the Handlebars prompt template factory to specify which helpers get registered and optionally inject custom helpers.

When registering helpers we should follow the same pattern as is used here https://github.com/Handlebars-Net/Handlebars.Net.Helpers. We don't want to put everything into a single class as indicated by using `private static` helper methods.

So there should be calls that look something like this:

```csharp
KernelSystemHelpers.Register(handlebarsContext);
KernelPromptHelpers.Register(handlebarsContext);
KernelPluginHelpers.Register(handlebarsContext, plugins);
```

Reply:
Added a `HandlebarsPromptTemplateOptions` that will allow users to filter by `KernelHelperCategories` as well as pass a callback to register their own helpers.